backend/app/controllers/spree/admin/stock_items_controller.rb

```diff
@@ -43,22 +43,29 @@ def load_stock_management_data
         @stock_locations = Spree::StockLocation.accessible_by(current_ability)
         @stock_item_stock_locations = params[:stock_location_id].present? ? @stock_locations.where(id: params[:stock_location_id]) : @stock_locations
         @variant_display_attributes = self.class.variant_display_attributes
-        @variants = Spree::Config.variant_search_class.new(params[:variant_search_term], scope: variant_scope).results
-        @variants = @variants.includes(:images, stock_items: :stock_location, product: :variant_images)
-        @variants = @variants.includes(option_values: :option_type)
-        @variants = @variants.order(id: :desc).page(params[:page]).per(params[:per_page] || Spree::Config[:orders_per_page])
+        @variants = Spree::Config.variant_search_class.new(params[:variant_search_term], scope: variant_scope).results.
+            order(id: :desc).page(params[:page]).per(params[:per_page] || Spree::Config[:orders_per_page])
       end
 
       def variant_scope
-        scope = Spree::Variant.accessible_by(current_ability)
-        scope = scope.where(product: @product) if @product
+        scope = Spree::Variant.accessible_by(current_ability, :read).includes(:images,
+                                                                              stock_items: :stock_location,
+                                                                              product: :variant_images,
+                                                                              option_values: :option_type)
+        scope.where!(product: @product) if @product
         scope = scope.order(:sku)
-        scope
+        add_stock_location_filtering(scope)
       end
 
       def collection
         []
       end
+
+      def add_stock_location_filtering(scope)
+        return scope if params[:stock_location_id].blank?
+
+        scope.joins(:stock_locations).where(spree_stock_locations: { id: params[:stock_location_id] })
```

Comment:
Perhaps we could move this to a scope? I usually prefer keeping query logic out of the controller, so we could add a scope like:

```ruby
scope :by_stock_location, -> (stock_location_id) {
  if stock_location_id
    joins(:stock_locations).where(spree_stock_locations: { id: params[:stock_location_id] })
  else
    all
  end
}
```

Then in the controller we'd just add `.by_stock_location(params[:stock_location_id].presence)` to the scope query.

Thoughts?

Reply:
Done! Thanks @jarednorman